### PR TITLE
MXNet distributed training

### PR DESCRIPTION
**Pull request by vandanavk**
Adding an example of distributed training using Apache MXNet.

---
This change is [Reviewable](https://reviewable.io/reviews/kubeflow/mpi-operator/122)

### CODE REVIEW

File: examples/mxnet/Dockerfile [hunk 1]
```diff
@@ -0,0 +1,27 @@
+FROM horovod/horovod:0.16.2-tf1.12.0-torch1.1.0-mxnet1.4.1-py3.5 AS build
```
Comment (rongou) [inline examples/mxnet/Dockerfile]: You don't need the `AS build` part.

File: examples/mxnet/Dockerfile [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+
+# Create a wrapper for OpenMPI to allow running as root by default
```
Comment (rongou) [inline examples/mxnet/Dockerfile]: These lines seem to have been removed in horovod's Dockerfile in favor of `horovodrun`. Have you tried to use it?

File: examples/mxnet/mxnet-mnist.yaml [hunk 3]
```diff
@@ -0,0 +1,50 @@
+apiVersion: kubeflow.org/v1alpha1
```
Comment (wackxu) [inline examples/mxnet/mxnet-mnist.yaml]: Could you add this example for v1alpha2 version?

**Comment by terrytangyuan** [comment]
Thank you @vandanavk!

File: examples/mxnet/mxnet_mnist.py [hunk 4]
```diff
@@ -0,0 +1,172 @@
...
+
+    # Evaluate model accuracy
+    _, train_acc = metric.get()
+    name, val_acc = evaluate(model, val_data, context)
+    if hvd.rank() == 0:
+        logging.info('Epoch[%d]\tTrain: %s=%f\tValidation: %s=%f', epoch, name,
+                     train_acc, name, val_acc)
+
+    if hvd.rank() == 0 and epoch == args.epochs - 1:
+        assert val_acc > 0.96, "Achieved accuracy (%f) is lower than expected\
```
Comment (terrytangyuan) [inline examples/mxnet/mxnet_mnist.py]: Avoid using `'\'`

File: examples/mxnet/mxnet_mnist.py [hunk 5]
```diff
@@ -0,0 +1,172 @@
...
+# Function to evaluate accuracy for a model
+def evaluate(model, data_iter, context):
+    data_iter.reset()
+    metric = mx.metric.Accuracy()
+    for _, batch in enumerate(data_iter):
+        data = batch.data[0].as_in_context(context)
+        label = batch.label[0].as_in_context(context)
+        output = model(data.astype(args.dtype, copy=False))
+        metric.update([label], [output])
+
```
Comment (terrytangyuan) [inline examples/mxnet/mxnet_mnist.py]: Remove this newline

File: examples/mxnet/mxnet-mnist.yaml [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+        - "1"
+        - --batch-size
+        - "64"
+        - --epochs
+        - "5"
+        image: mpioperator/mxnet-horovod:latest
+        name: mxnet-mnist-horovod-job
+        resources:
+          limits:
+            nvidia.com/gpu: 4
```
Comment (terrytangyuan) [inline examples/mxnet/mxnet-mnist.yaml]: Add newline

File: examples/mxnet/mxnet-mnist.yaml [hunk 7]
```diff
@@ -0,0 +1,50 @@
+apiVersion: kubeflow.org/v1alpha2
```
Comment (terrytangyuan) [inline examples/mxnet/mxnet-mnist.yaml]: Have you verified that this is working? The spec still looks like v1alpha1.

File: examples/mxnet/mxnet_mnist.py [hunk 8]
```diff
@@ -0,0 +1,172 @@
...
+import zipfile
+import time
+
+import mxnet as mx
+import horovod.mxnet as hvd
+from mxnet import autograd, gluon, nd
+from mxnet.test_utils import download
+
+# Training settings
+parser = argparse.ArgumentParser(description='MXNet MNIST Example')
```
Comment (terrytangyuan) [inline examples/mxnet/mxnet_mnist.py]: Could you add "Apache"?

**Comment by vandanavk** [comment]
> Thanks for your pull request. It looks like this may be your first contribution to a Google open source project (if not, look below for help). Before we can look at your pull request, you'll need to sign a Contributor License Agreement (CLA).
> 
> 📝 **Please visit https://cla.developers.google.com/ to sign.**
> 
> Once you've signed (or fixed any issues), please reply here (e.g. `I signed it!`) and we'll verify it.
> 
> #### What to do if you already signed the CLA
> ##### Individual signers
> * It's possible we don't have your GitHub username or you're using a different email address on your commit. Check [your existing CLA data](https://cla.developers.google.com/clas) and verify that your [email is set on your git commits](https://help.github.com/articles/setting-your-email-in-git/).
> 
> ##### Corporate signers
> * Your company has a Point of Contact who decides which employees are authorized to participate. Ask your POC to be added to the group of authorized contributors. If you don't know who your Point of Contact is, direct the Google project maintainer to [go/cla#troubleshoot](http://go/cla#troubleshoot) ([Public version](https://opensource.google.com/docs/cla/#troubleshoot)).
> * The email used to register you as an authorized contributor must be the email used for the Git commit. Check [your existing CLA data](https://cla.developers.google.com/clas) and verify that your [email is set on your git commits](https://help.github.com/articles/setting-your-email-in-git/).
> * The email used to register you as an authorized contributor must also be [attached to your GitHub account](https://github.com/settings/emails).
> 
> ℹ️ **Googlers: [Go here](https://goto.google.com/prinfo/https%3A%2F%2Fgithub.com%2Fkubeflow%2Fmpi-operator%2Fpull%2F122) for more info**.

I signed it!

File: examples/mxnet/Dockerfile (hunk 2 above)
Comment (vandanavk) [reply examples/mxnet/Dockerfile]: I haven't tried horovodrun. Should this example be using horovodrun?

Comment (rongou) [reply examples/mxnet/Dockerfile]: I guess it's up to you. We can certainly do it at a later time.

**Pull request merged**